### ISSUE TITLE
Cast operation failed after a complex loop

### ISSUE THREAD

**Issue by JavaTailor** [eclipse-openj9/openj9#14716]
## Affected versions

We found a test case with execution problems. To facilitate analysis, we simplified the test case and the simplified class file can be found at [attachment](##Attachment).

Windows 10:

```
Microsoft Windows 10 Professional
10.0.19044 Build 19044
AMD Ryzen 5 5600G with Radeon Graphics 3.90 GHz
Memory 32 GB
```

Java -version output under Windows 10

```
openjdk version "1.8.0_312"
IBM Semeru Runtime Open Edition (build 1.8.0_312-b07)
Eclipse OpenJ9 VM (build openj9-0.29.0, JRE 1.8.0 Windows 10 amd64-64-Bit Compressed References 20211022_236 (JIT enabled, AOT enabled)
OpenJ9   - e1e72c497
OMR      - 299b6a2d2
JCL      - 8860d39588 based on jdk8u312-b07
```

```
openjdk version "11.0.13" 2021-10-19
IBM Semeru Runtime Open Edition 11.0.13.0 (build 11.0.13+8)
Eclipse OpenJ9 VM 11.0.13.0 (build openj9-0.29.0, JRE 11 Windows 10 amd64-64-Bit Compressed References 20211022_218 (JIT enabled, AOT enabled)
OpenJ9   - e1e72c497
OMR      - 299b6a2d2
JCL      - 2d83aa3b76 based on jdk-11.0.13+8)
```

```
openjdk version "16.0.2" 2021-07-20
IBM Semeru Runtime Open Edition 16.0.2.0 (build 16.0.2+7)
Eclipse OpenJ9 VM 16.0.2.0 (build openj9-0.27.0, JRE 16 Windows 10 amd64-64-Bit Compressed References 20210730_69 (JIT enabled, AOT enabled)
OpenJ9   - 1851b0074
OMR      - 9db1c870d
JCL      - 34df42439f3 based on jdk-16.0.2+7)
```

## Problem summary

The decompiled code of the test case is shown below, it started by building an ArrayList object var5. And there are two cast operations which cast var5 to List at Point 1 and Point 2 respectively, however, the cast results inside and outside the loop are different. The cast was failed at Point 2, and thus we got the "failure2" printed.

```java
ArrayList var5 = new ArrayList();

for(int var2 = 0; var2 < 5000; ++var2) {
    ArrayList var1 = new ArrayList();
    for(int var3 = var2; var3 < var2 + 1000; ++var3) {
        TestCase1 var4 = new TestCase1();
        var1.add(var4);
    }
    var5.add(var1);
    if ((List)var5 == null) { // Point 1
        System.out.println("failure1"); 
    }
}

if ((List)var5 == null) { // Point 2
    System.out.println("failure2"); 
} else {
    System.out.println("success");
}
```

Moreover, we found that there is an similar issue #12819 submitted before. However, the problems of these two issues are similar but not the same. The issue in #12819 was that the cast operation inside the loop caused subsequent bytecode to be optimized, whereas this issue occurred outside of the loop, that is, the cast operation at Point 2 produces an error. The issue in #12819 has been fixed, but the fixed version continued to execute abnormally on this test case.

## Attachment
[TestCase1.zip](https://github.com/eclipse-openj9/openj9/files/8250000/TestCase1.zip)

**Comment by pshipton** [comment]
Passes with `-Xint` and `-Xjit:optlevel=cold`, fails with `-Xjit:optlevel=warm`. Added to 0.32 milestone for evaluation by JIT.

**Comment by 0xdaryl** [comment]
Possibly a DLT issue.  Only fails in the DLT compile.

**Comment by 0xdaryl** [comment]
A quick update on this.  This seems to be tangentially related to 12819, but a different problem.  The parameter slot (declared a `java/lang/String`) is indeed shared with an `ArrayList` in the bytecode.  This slot is updated early in the method upon allocating a new `ArrayList`.  However, when the method is DLT'd, the DLT entry point to the method occurs after that point and the IL prior to the DLT entry point is eliminated.  Because it has been eliminated, the store to the parm slot is not seen by GVP and hence it concludes (arguably correctly given the IL it has to work with) that the parm slot must be a `java/lang/String` type because nothing was ever stored to it.  This further means that the checkcasts in the method will fail because a String is not a List, and the code is folded assuming so.

I haven't decided on the best way to fix this yet.

**Comment by 0xdaryl** [comment]
Proposed fix in eclipse/omr#6423.

**Comment by 0xdaryl** [comment]
The upstream PR eclipse/omr#6423 has been merged.

**Comment by pshipton** [comment]
Fixed in the 0.32 branch.
